### ISSUE TITLE
Early exit RSS feed processing once we see old content

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#2998]
The current algorithm we use to process the RSS feeds is overly thorough and the expense of making hundreds of unnecessary queries. 

Right now, when we detect a new RSS feed, we:

1. Iterate over each item in the feed.
2. We make a hash of each item.
3. We check our cache for that hash.
4. If it's an item we've seen before, we continue to the next.
5. If it's something new, we:
    - add it to the DB, and 
    - cache the hash for next time.

We should add an early exit after step 4, since the feeds are chronological (they *are*, right?). If that's too aggressive, (@johnhawkinson I suspect you'll have opinions?), perhaps we stop parsing the feed after hitting the cache X times?

Some of these feeds are very big, and it's just a waste of resources to do the entire feed every time it says it has changed. I'm thinking about this because I'm watching celery chug through these, on a weekend, and that makes very little sense.

**Comment by mlissner** [comment]
John did have a good suggestion. Instead of exiting after hitting the cache too many times, we can instead compare the feed date with the item date. So if the feed was updated at midnight, any items in the feed with earlier dates can be ignored. That'll be fast enough and more foolproof than my suggestion.

As a benchmark, we just did a feed with 1300 items and it took 80s.

**Comment by albertisfu** [comment]
> John did have a good suggestion. Instead of exiting after hitting the cache too many times, we can instead compare the feed date with the item date. So if the feed was updated at midnight, any items in the feed with earlier dates can be ignored. That'll be fast enough and more foolproof than my suggestion.


One question here @mlissner to confirm I understand the approach correctly.

If we find a cached item, that means we’ve already processed at least some items in the feed. So, once we encounter a cached item, instead of checking whether the following items are also cached to ignore them, we should check each item's date and ignore it if it’s earlier than the feed’s build time. If we found an item that it's later than the build time we do the regular process, check if it's cached, process it and finally cache it. Is that correct?

The idea behind this is that some feeds might not be in chronological order, so this approach would be more robust?

**Comment by mlissner** [comment]
Exactly! This should be a good way to be sure everything is ingested without making 1300 queries to the DB.

**Comment by johnhawkinson** [comment]
Of course, we should be aware that the problem with that is that history can change. E.g.:

10:01am, SDNY dockets entry Alpha
11:00am, SDNY dockets entry Beta
11:05am, SDNY generates the every 2-hour RSS feed
12:20pm ,SDNY clerk edits the entry to add/change/whatever
13:00pm, SDNY dockets entry Gamma
13:05pm, SDNY generates the every 2-hour RSS feed

If we process the feed at 11:30am and get the 11:05am version and process Alpha and Beta and then we process the feed at 12:30pm and it starts with Beta so we ignore Beta and Alpha an then we process the feed at 1:30pm and it starts with Gamma which is to new to us, so we process Gamma, and then we see Beta and it is cached so we stop.

Well, we'll never see that Alpha has changed.

I think, fundamentally, I don't understand why this is so expensive to cache and compare. It seems like it should not be.

I don't mean to say this happens a lot or is a huge deal, but it is a limitation to be aware of.

**Comment by mlissner** [comment]
> I think, fundamentally, I don't understand why this is so expensive to cache and compare. It seems like it should not be.

There's something like 100 courts with RSS feeds that contain upwards of 1,000 items each, so that means doing 100k cache checks each time we scrape. That's a lot of DB work. It's fast DB work, but it's a lot of DB work. (I measured 80s for 1300 items above.)

**Comment by johnhawkinson** [comment]
214 courts, 1,000 items per, so 214,000 cache entries.
That is not big. That is not a lot.
A data structure with 214,000 hashes should be near-instant to query?
Maybe it's in an SQL DB when it should be an RAM or in a disk-backed cache on a single machine.
I dunno.

It just seems off to me.

**Comment by mlissner** [comment]
Yeah, moving it out of the DB might be better. Maybe in the next iteration, unless Alberto gets a wild hair.

**Comment by albertisfu** [comment]
> If we process the feed at 11:30am and get the 11:05am version and process Alpha and Beta and then we process the feed at 12:30pm and it starts with Beta so we ignore Beta and Alpha an then we process the feed at 1:30pm and it starts with Gamma which is to new to us, so we process Gamma, and then we see Beta and it is cached so we stop.
> 
> Well, we'll never see that Alpha has changed.

Well, the approach we're planning to apply here is to compare each item's `pubDate` with the  feed `lastBuildDate`. We'll perform this comparison for all items instead of relying on the cache. This way, we'll be able to identify which items need to be processed, specifically, those with a `pubDate` later than the `lastBuildDate`.

The question is: if an item changes (like Alpha in your example), will its `pubDate` be updated? or will it remain the same as when it was originally published? If it's the latter, I think this approach might miss those changes.

**Comment by mlissner** [comment]
> The question is: if an item changes (like Alpha in your example), will its pubDate be updated? or will it remain the same as when it was originally published? If it's the latter, I think this approach might miss those changes.

We are missing these now too, aren't we?

**Comment by albertisfu** [comment]
> We are missing these now too, aren't we?

Well, if the item changes anything in its content, the hash changes too. So currently, we detect that the item is not cached and ingest it.

**Comment by johnhawkinson** [comment]
> The question is: if an item changes (like Alpha in your example), will its `pubDate` be updated? or will it remain the same as when it was originally published? If it's the latter, I think this approach might miss those changes.

I assume the `pubDate` won't change, but I don't know. Since you have the corpus of raw rss feeds, should be easy enough to check.

> Well, if the item changes anything in its content, the hash changes too. So currently, we detect that the item is not cached and ingest it.

Indeed, the whole point of the hash ;)

**Comment by albertisfu** [comment]
Regarding this issue, I’ve been working on confirming this question:

> I assume the pubDate won't change, but I don't know. Since you have the corpus of raw rss feeds, should be easy enough to check.

If this is true, then the approach described in the issue wouldn’t be suitable, since it could lead to missing entry updates that currently are being captured.

To test this, I wrote a script that explores the `RssFeedData` objects from a court across multiple days and times of day (three different times). The script identifies entries by `pacer_doc_id`, hashes each entry, and compares the hashes for the same `pacer_doc_id` across different times of the day.

Finally, the script reports any `pacer_doc_ids` where the hashes differ, meaning the content of the entry has changed. This allows us to manually check the RSS feed, confirm what changed, and determine whether the pubDate changed as well.

Here is the script:

[scratch_154.py](https://github.com/user-attachments/files/22675897/scratch_154.py)

I ran for `txwd` and `cacd`, and it wasn’t difficult to find entries that were updated on the same day.

The bad news is that when an entry is updated, the `pubDate` is not updated. It remains the original one.

**TXWD**
**<lastBuildDate>Tue, 30 Sep 2025 15:01:02 GMT</lastBuildDate>**
```
<item>
<title>1:25-cv-00736 Garner et al v. Travis County Emergency Services District No. 1</title>
<link>https://ecf.txwd.uscourts.gov/cgi-bin/DktRpt.pl?1172842425</link>
<description>[Supplement to] (&#x3C;a href=&#x22;https://ecf.txwd.uscourts.gov/doc1/181035059646?caseid=1172842425&#x26;de_seq_num=63&#x22; &#x3E;19&#x3C;/a&#x3E;)</description>
<guid isPermaLink="true">https://ecf.txwd.uscourts.gov/cgi-bin/DktRpt.pl?1172842425&#x26;63</guid>
<pubDate>Tue, 30 Sep 2025 14:02:15 GMT</pubDate>
</item>
```
[637d8a21a59842309f67cebc76cc3375.zip](https://github.com/user-attachments/files/22675976/637d8a21a59842309f67cebc76cc3375.zip)

**<lastBuildDate>Wed, 01 Oct 2025 06:01:02 GMT</lastBuildDate>**
```
<item>
<title>1:25-cv-00736 Garner et al v. Travis County Emergency Services District No. 1</title>
<link>https://ecf.txwd.uscourts.gov/cgi-bin/DktRpt.pl?1172842425</link>
<description>[Miscellaneous Relief] (&#x3C;a href=&#x22;https://ecf.txwd.uscourts.gov/doc1/181035059646?caseid=1172842425&#x26;de_seq_num=63&#x22; &#x3E;19&#x3C;/a&#x3E;)</description>
<guid isPermaLink="true">https://ecf.txwd.uscourts.gov/cgi-bin/DktRpt.pl?1172842425&#x26;63</guid>
<pubDate>Tue, 30 Sep 2025 14:02:15 GMT</pubDate>
</item>
```
[fb6465028ec940589d1b9e6053bb4638.zip](https://github.com/user-attachments/files/22675973/fb6465028ec940589d1b9e6053bb4638.zip)

You can observe that the entry description changed from `Supplement to` to `Miscellaneous Relief`, but the `pubDate` remained the same.


**CACD**
**<lastBuildDate>Thu, 02 Oct 2025 15:00:01 GMT</lastBuildDate>**
```
<item>
<title>2:25-cv-03791 Darian Nikia Munson v. Covarrubias et al</title>
<link>https://ecf.cacd.uscourts.gov/cgi-bin/DktRpt.pl?969045</link>
<description>[Notice (Other)] (&#x3C;a href=&#x22;https://ecf.cacd.uscourts.gov/doc1/031145987453?caseid=969045&#x26;de_seq_num=67&#x22; &#x3E;23&#x3C;/a&#x3E;)</description>
<guid isPermaLink="true">https://ecf.cacd.uscourts.gov/cgi-bin/DktRpt.pl?969045&#x26;67</guid>
<pubDate>Thu, 02 Oct 2025 14:16:47 GMT</pubDate>
</item>
```
[0203655325ad4be6b30420b93a555ccc.zip](https://github.com/user-attachments/files/22676017/0203655325ad4be6b30420b93a555ccc.zip)

**<lastBuildDate>Fri, 03 Oct 2025 06:00:02 GMT</lastBuildDate>**
```
<item>
<title>2:25-cv-03791 Darian Nikia Munson v. Covarrubias et al</title>
<link>https://ecf.cacd.uscourts.gov/cgi-bin/DktRpt.pl?969045</link>
<description>[Order] (&#x3C;a href=&#x22;https://ecf.cacd.uscourts.gov/doc1/031145987453?caseid=969045&#x26;de_seq_num=67&#x22; &#x3E;23&#x3C;/a&#x3E;)</description>
<guid isPermaLink="true">https://ecf.cacd.uscourts.gov/cgi-bin/DktRpt.pl?969045&#x26;67</guid>
<pubDate>Thu, 02 Oct 2025 14:16:47 GMT</pubDate>
</item>
```
[8700ad43b4d54b4597315a2f9bfa0dbf.zip](https://github.com/user-attachments/files/22676007/8700ad43b4d54b4597315a2f9bfa0dbf.zip)

The description changed from `Notice (Other)` to `Order,` while the `pubDate` again remained unchanged.

In conclusion. It is both possible and common for entries to change during the same day while the `pubDate` stays the same. Therefore, if we apply the approach described in this issue, aborting feed processing once we encounter an RSS entry older than the `lastBuildDate`, we risk missing updates like these.

Let me know what do you think @mlissner

**Comment by mlissner** [comment]
OK, so that's good to know and be certain of. 

So I think the next best idea was to store these hashes in memory instead of in the DB so that it's 200k in-memory look-ups instead of 200k DB queries. Do you have thoughts about how hard and reliable that'd be, @albertisfu?

**Comment by albertisfu** [comment]
> So I think the next best idea was to store these hashes in memory instead of in the DB so that it's 200k in-memory look-ups instead of 200k DB queries. Do you have thoughts about how hard and reliable that'd be, @albertisfu?

Yeah! That’s 200k hashes in memory. Since our hashes are SHA-256 (32 bytes), we’re talking about ~7MB of storage, which seems safe to keep in Redis. The query should also be very fast. So yeah, this sounds like the right approach.

 I’ll work on moving to this approach then.
